Use file storage as cache backing store

tsunyoku opened the pull request:
Resolves #1.

This creates a new `IReplayCache` abstraction in case the backing store ever changes, but right now there is only `FileReplayCache`.

There are new environment variables to set the storage path for the cache. This is intentionally different from the local storage path as they should always be separated but in production the local storage paths are not used anyways. This should allow for block volumes to be used without issue.

This cache works by creating folders for each day to contain cached replays in, with a worker in the background to delete these folders based on age once they hit the maximum specified by the new env var for it.

tsunyoku committed:
Use file storage as backing store for cache

tsunyoku committed:
Fix incorrect indexes

tsunyoku committed:
Prevent Redis subscriber from swallowing exceptions

tsunyoku committed:
Handle `RedisChannel` warning

tsunyoku committed:
Simplify `RedisDatabase` construction

tsunyoku committed:
Add new environment variables to README

tsunyoku committed:
Remove unused cache dependency

tsunyoku committed:
Update tests

bdach commented on osu.Server.ReplayStore/Services/FileReplayCache.cs:
```diff
@@ -0,0 +1,73 @@
+
+            await db.StringSetAsync(
+                getCacheKey(scoreId, rulesetId, legacyScore),
+                1, // filler value
```
Not sure how to feel about this type of thing. Involving redis here at all is a bit of a much of muchness altogether to me if I'm honest. I know it was probably because of [this comment](https://github.com/ppy/osu-server-replay-store/issues/1#issuecomment-3227502154) but I dunno. It makes management of everything so much messier for very little benefit.

Now my thoughts are darting all over about what happens if a redis operation succeeds but the disk IO fails, or the converse, or about data races, and all of this... to basically have a persistent hashset with entry expiry... for a service that is going to be single-instanced? I dunno. It just feels like shooting a fly with a bazooka.

I'm also a little worried about this just putting keys over the "default" redis database with no real separation or namespacing. Is this intended to run on the shared redis instance, or its own one?

tsunyoku replied:
What would you suggest as an alternative for tracking expiries then? I'm not married to the idea of using Redis other than it being a common choice for this kind of thing

bdach replied:
Simpler things - to me - would be:

- Just using the filesystem. One folder per day. Use a partitioning scheme wherein you dump the replays to a dated folder, and on date rollover, make a new folder and delete the old one.
- Using filesystem metadata (`mtime`). May be slow to scan for expiry, not sure about that one.
- Just construct an in-memory store locally to dotnet and use that. It'll be very similar to what you have with redis, except 100% less network calls, and with 100% less worry about desync between redis and disk (you'd either reconstruct the mapping on restart, or just purge disk).

That said before you do anything with any of the above I'd like to rack @peppy's brains on all this.

tsunyoku replied:
Personally think the first option is best. Second is probably slow, and third is even *less* fool proof than Redis since all context on expiries will die with service restarts etc.

I'll wait for peppy's comment but if we're replacing Redis then I think filesystem is the only way forward for it to be any improvement.

bdach replied:
> and third is even _less_ fool proof than Redis since all context on expiries will die with service restarts etc.

Not necessarily, as I said, you could use `mtime` of the files to reconstruct the expiries. As far as I am aware, the design parameters put down here are such that the *service* itself is not supposed to be reliable or redundant or whatever. The *upload primitive* must be resilient via *some* means, but the download primitive? Not particularly.

I don't like redis because it introduces a possibility of a split-brain where the state of the redis expiry tracking no longer matches the state of the disk cache. I would like that possibility to not be there (or at least be reduced).

tsunyoku replied:
I understand the concern about Redis and disk desync, but equally I'd like to avoid complexity where I can. I think the first suggestion is the simplest option that doesn't introduce new problems.

peppy replied:
I think filesystem cycling per day sounds fine, although I'd be interested in how the lookup logic for that works (just check in each subfolder for simplicity? or try and make the date of the replay correctly match for a single query?)

---

**Ignore the following, I wrote before checking actual S3 usage and seeing how little we spend on replay bandwidth**

Thing is, I do see a world where we don't want to cycle per day (which would mean moving files across different filesystem folders to "bump" their last access time).

If we're using filesystem, I'd much rather have it all in a permanent folder structure or even just using a local solution like ceph (but hopefully something simpler and more modern) which may come with it's own LRU tracking.

Given that block storage is relatively cheap ($150/tb) i would see us tuning things to cache a large amount locally to reduce s3 throughput, rather than churning the cache quickly.

bdach replied:
> I'd be interested in how the lookup logic for that works

I was imagining that the folder would be based on date of population of cache rather than date of when the score was set. So the lookup logic would be literally just "only check today's folder as dictated by the server clock".

I dunno. I just really don't see much of a point in involving redis in a way like this (that I consider rather hacky). It's making TCP calls over network and opening up for split-brain desync badness just for *cache expiry tracking* for a *non-essential service*...

peppy replied:
My concern is the hard cutoff at "midnight". So I'd probably see us querying each of the partitioned folders. Probably fine to do this overhead wise if it's local / block storage?

bdach replied:
Yeah I can see the midnight argument. In that case maybe you could do something like a 2-day cycle wherein a folder only gets purged after 2 days.

> Probably fine to do this overhead wise if it's local / block storage?

I'm not that familiar with latency numbers for block storage, but unless this service gets to live on a box with a dedicated redis service on localhost, I think there's a 99.999% chance that even querying each folder is going to be faster than having to put a bunch of ones in redis.

peppy replied:
Block storage is still in the milliseconds at worst (using DO's which mounts as a filesystem).

bdach commented on osu.Server.ReplayStore/Services/FileReplayCache.cs:
```diff
@@ -0,0 +1,73 @@
+            await db.KeyDeleteAsync(
+                getCacheKey(scoreId, rulesetId, legacyScore));
+
+            File.Delete(getPathToReplay(scoreId, rulesetId, legacyScore));
```
I sure hope that this is going to be running on a unix or another system that has the nice `unlink(2)` semantics, because if this ever runs on a windows box, then I can already see this failing over when a replay deletion is requested while another reader is still not done reading it.

At least the ordering of redis-before-IO here looks safer than the contrary (still risks leaving dead replays that will never be evicted on disk, but that's better than the alternative)

tsunyoku replied:
Not really sure what you want me to do about this

bdach replied:
Nothing, just a passing comment. As things are not much *can* be done about this except not run on windows, because locking here would be a *baaaad* idea (and also probably literally impossible to do).

The outlook may change as the discussion on using redis at all progresses.

bdach commented on osu.Server.ReplayStore/ExpireReplayCacheWorker.cs:
Where do the "keyspace events" go if the service happens to be down while redis emits them? I suppose the answer to that is `/dev/null`?

If so, then that's another point of worry in involving redis in any of this to me.

bdach commented on osu.Server.ReplayStore/ExpireReplayCacheWorker.cs:
```diff
@@ -0,0 +1,97 @@
+                // Due to this method being called by a Redis subscriber, exceptions fail to be raised properly.
+                // Log the error and capture it to Sentry so that we don't ignore them.
+                logger.LogError(ex, "Failed to process keyspace event");
+                SentrySdk.CaptureException(ex);
```
The capture to sentry seems redundant. Sentry is already plugged into logging in

https://github.com/ppy/osu-server-replay-store/blob/77f1feefb05172fb51d9e4cf17e2fbd91f4b31f1/osu.Server.ReplayStore/Program.cs#L36-L37

and so I'm pretty sure calling `logger.LogError()` will already send it to sentry.

bdach reviewed (commented):
Sorry for leaving this unreviewed for as long as it was. That said I'm not sure you're going to be happy with my review either...

I have concerns with leaning on redis again for the expiry tracking. Especially given this:

> The eviction works by setting a dummy key on Redis with an expiry, and then we capture the expiry using Redis' [keyspace events](https://redis.io/docs/latest/develop/pubsub/keyspace-notifications/) in the background. In order for these events to be captured, you will need to run the following command on the indicated Redis instance to capture expire(d) and eviction events:
> 
> ```
> config set notify-keyspace-events K$xe
> ```

as I worry about the performance implications of doing this if this is ran on a redis instance that is not completely dedicated to this service.

@peppy probably have a read through this and see if you agree with my points, since most of them are subjective architectural "feels" rather than concrete criticisms.

tsunyoku committed:
Replace Redis solution with a caching folders per day

tsunyoku commented on osu.Server.ReplayStore/ReplayStoreController.cs:
```diff
@@ -55,15 +53,10 @@ public async Task<IActionResult> PutReplayAsync(
             byte[] replayBytes = await replayStream.ReadAllRemainingBytesToArrayAsync();
 
-            await replayStorage.StoreReplayAsync(scoreId, score.ruleset_id, legacyScore: false, replayStream);
+            replayStream.Seek(0, SeekOrigin.Begin);
```
This and other new calls to `replayStream.Seek` are a bug fix. Apparently I somehow didn't notice non-cached replays saving as empty files (because the stream had already been seeked to the end by creating `replayBytes` just above)

bdach replied:
That is *catastrophic* failure.

I would expect bug fixes of this magnitude to have test coverage. The changes done to the tests here look insufficient to cover this (they appear to be mostly checking if an byte array is null or not, and not checking if it's zero bytes)

tsunyoku committed:
Fix test passing incorrect cache folder

tsunyoku committed:
Fix replay cache folder for non-legacy case in tests too

tsunyoku committed:
Fix weird test method names

tsunyoku committed:
Fix incorrect method call for cache directory

tsunyoku committed:
Create unique temp path for each test run

Required to ensure cache folders are unique per test-run

tsunyoku commented:
As requested, this no longer uses Redis to keep track of the cache. As a result there are no mentions of Redis _at all_ in this project now, although that will change once I PR a solution for #2.

This works by creating subdirectories for each date (at time of upload), with a worker that will constantly check each subdirectory and delete it if it passes the maximum amount of cached days (as per app settings).

tsunyoku committed:
Fix cache lookup failing for replays cached prior to today

bdach commented on osu.Server.ReplayStore/Services/FileReplayCache.cs:
Class should contain xmldoc which spells out what the directory structure is. Right now it has to be pieced together from circumstantial pieces.

bdach commented on osu.Server.ReplayStore/Services/FileReplayCache.cs:
```diff
@@ -0,0 +1,84 @@
+        public async Task<byte[]?> FindReplayDataAsync(long scoreId, ushort rulesetId, bool legacyScore)
+        {
+            string baseCacheDirectory = legacyScore
+                ? string.Format(legacyBaseDirectory, LegacyRulesetHelper.GetRulesetNameFromLegacyId(rulesetId))
```
This expects `legacyBaseDirectory` to contain a string format placeholder for the ruleset name. But if it doesn't have one, this will fail catastrophically with no warning (either by crashing on things, or worse, by mixing replays from other rulesets, overwriting some with others, etc.) I definitely wouldn't expect having to put in a string format placeholder for this envvar.

I'm not sure why the string format placeholder is supposed to be part of the path here anyway. Just manage the structure directly. `legacyBaseDirectory` points to the root, everything below is hardcoded.

tsunyoku replied:
`LocalReplayStorage` is also doing this - I'll change it in both places

bdach commented on osu.Server.ReplayStore/ExpireReplayCacheWorker.cs:
```diff
@@ -0,0 +1,54 @@
+                    Directory.Delete(cacheFolder, true);
+                    logger.LogInformation("Deleted expired cache folder {CacheFolder}", cacheFolder);
+                }
+            }
```
No try-catches, no logging of failures? Can't do that on a long-running service or it'll die on the first failure, thus silently stop purging cache and then explode from running out of disk.

tsunyoku committed:
Add test cases for incorrect stream handling

tsunyoku committed:
Add try-catch and error logging for expiry worker

tsunyoku committed:
Split legacy ruleset replays into subfolders instead of format-based app setting

tsunyoku committed:
Add xmldoc to explain cache folder structure

tsunyoku committed:
Move around test setup

bdach committed:
Actually inspect stored replay

The bug fix mentioned in
https://github.com/ppy/osu-server-replay-store/pull/12#discussion_r2441208929
was attempted to be covered with tests added in
18309f244c3462a7e5deac1a1a69b63fa59a49b2, but those test changes *don't
actually cover the failure*, because they exercised the *cached* replay,
which wasn't affected by the lack of seek, instead of the *stored*
replay, which *was* affected by the lack of seek.

bdach committed:
Improve xmldoc

bdach reviewed (approved):
Probably fine.

@tsunyoku please check my latest commits, https://github.com/ppy/osu-server-replay-store/pull/12/commits/b021ec7daf70f831ce3fa7ad1977a529d4578c9a in particular.

@peppy not sure if you want to have a second pair of eyes on this.

tsunyoku commented:
Looks good.

tsunyoku commented:
Are we good to merge this? I'm gonna need this in to work on #2.

peppy commented:
Checking on this today.

peppy commented on osu.Server.ReplayStore/Services/FileReplayCache.cs:
```diff
@@ -0,0 +1,121 @@
+    /// In the case of the legacy cache folder, replays must be split by ruleset, because stable scores have separate ID schemes per ruleset,
+    /// so there is another hierarchy level inside with a folder per ruleset.
+    /// At the lowest level, the cache is divided up by folders of each date.
+    /// When a replay is added to the cache, it will be put into a folder named by the date it was added in <c>ddMMyy</c> format.
```
Is there a valid reason we're splitting out legacy vs non-legacy?

tsunyoku replied:
Mostly because thats how it works currently. Does web-10 have the guarantee that it will know the non-legacy ID at point of uploading the replay? What about API v1 or legacy scores via API v2 that request the replay using the legacy score ID?

tsunyoku replied:
Also currently the upload flow relies on separating legacy and non-legacy uploads since non-legacy uploads have replay headers attached and non-legacy ones don't. I'm happy to move things to a more unified model if we have the guarantees questioned in my other comment, but it will be an undertaking in itself.

peppy replied:
maybe the initial upload (if we're actually planning to do that from web-10 and not make it in new code) may have issues. the other uses you mention won't (we maintain a mapping).

also don't expect legacy score tables to be a thing for too much longer. we have short-term plans of nuking it.

tsunyoku replied:
Okay cool. I'll think of a way to unify the replay process for both clients then. It's likely just going to be checking for `legacy_score_id` on the `scores` table unless you have a preferred method of doing this (I still need to know if a score is set in stable so I can handle replay headers)

tsunyoku replied:
I've done this the best I can in 2ed5846a335583cf8e544bdd4ec4818225bfa9e9. Please read the commit description for a better understanding of what legacy parts remain.

As far as I'm aware, keeping the *storage* split between legacy and solo is a non-negotiable given all legacy replays are currently stored in the legacy replay buckets and unifying this so that even legacy replays upload to the new solo bucket would require logic to check all buckets when trying to fetch a replay for a legacy score (since there's the chance it could be in any of the buckets). If we feel strongly about that, I can implement the logic to check each bucket and remove that one remaining legacy distinction but I'm not immediately convinced its worth it.

This still feels a lot better though, since there is now one set of endpoints for _all_ replays

peppy commented on osu.Server.ReplayStore/Services/FileReplayCache.cs:
```diff
@@ -0,0 +1,121 @@
+
+        private string getReplayDirectory(ushort rulesetId, bool legacyScore)
+        {
+            string date = DateTime.Today.ToString("ddMMyy");
```
yyyyMMdd please

peppy commented on osu.Server.ReplayStore/Services/FileReplayCache.cs:
```diff
@@ -0,0 +1,121 @@
+    ///  
+    /// $(AppSettings.LegacyReplayCacheStoragePath)
+    /// ├─ osu
+    /// │  ├─ 301025
```
is there a reason the dates are underneath ruleset rather than vice versa? feels back to front. cleanup sounds like a PITA.

peppy commented on osu.Server.ReplayStore/ExpireReplayCacheWorker.cs:
```diff
@@ -0,0 +1,64 @@
+                }
+                finally
+                {
+                    await Task.Delay(TimeSpan.FromHours(1), stoppingToken);
```
Is chaining like this fine these days? It used to cause call stack tail issues.

bdach replied:
On a quick check

```csharp
int stackDepth = 0;
Recurse();

void Recurse()
{
    if (Interlocked.Increment(ref stackDepth) % 1000 == 0)
        Console.WriteLine(stackDepth);
    Recurse();
}

```

dies at ~261k stack depth, while

```csharp
var cts = new CancellationTokenSource();

Console.CancelKeyPress += (_, _) => cts.Cancel();
int stackDepth = 0;

while (!cts.IsCancellationRequested)
{
    if (Interlocked.Increment(ref stackDepth) % 1000 == 0)
        Console.WriteLine(stackDepth);
    await Task.Delay(1);
}
```

continues spinning past 300k. IL inspection indicates that maybe C# is smart enough to unroll this pattern into a goto.

peppy replied:
That's good to know. It wasn't previously (we've had this exact loop fail in the wild).

tsunyoku committed:
Use `yyyyMMdd` format for cache directory

tsunyoku committed:
Restructure cache for date to be the higher level folder

tsunyoku committed:
Remove as much legacy specialities as possible

This attempts to unify legacy and solo replays as much as possible. All endpoints now take solo score IDs only. Long-term replay storage (usually S3) still splits between legacy and solo to support existing systems, but all caching is unified. All references to legacy score tables are nuked.

tsunyoku committed:
Fix binary types on legacy replay encoding

tsunyoku committed:
Make sure non-osu rulesets set 100s, gekis and katus correctly

Referenced from https://github.com/ppy/osu-web/blob/a007d8541869b67962735877a0fa35f3201a66f6/app/Models/Solo/Score.php#L458-L481

tsunyoku committed:
Consolidate all replay storage into a single bucket

After discussion, it appears the plan is likely to migrate legacy replays into a single bucket with their headers appended.

tsunyoku committed:
Formatting
